src/python/pants/backend/native/tasks/native_compile.py

```diff
@@ -0,0 +1,248 @@
+# coding=utf-8
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+import os
+from abc import abstractmethod
+from collections import defaultdict
+
+from pants.backend.native.config.environment import Executable
+from pants.backend.native.targets.native_library import NativeLibrary
+from pants.backend.native.tasks.native_task import NativeTask
+from pants.base.exceptions import TaskError
+from pants.build_graph.dependency_context import DependencyContext
+from pants.util.memo import memoized_method, memoized_property
+from pants.util.objects import SubclassesOf, datatype
+
+
+class NativeCompileRequest(datatype([
+    ('compiler', SubclassesOf(Executable)),
+    # TODO: add type checking for Collection.of(<type>)!
+    'include_dirs',
+    'sources',
+    ('fatal_warnings', bool),
+    'output_dir',
+])): pass
+
+
+# TODO: verify that filenames are valid fileNAMES and not deeper paths? does this matter?
```

Comment:
Would likely be better to use `Snapshot`s moving forward, which handles this for you. But that's a followup.

Reply:
Added link to #5950.